pymtl3/datatypes/bit_structs.py

```diff
@@ -0,0 +1,476 @@
+# function that looks like the following:
+#
+# def __init__( s, x=_type_x(), y=_type_y() ):
+#   s.x = x
+#   s.y = y
+#
+# NOTE:
+# _mk_init_fn also takes as argument the name of self in case there is a
+# field with name 's' or 'self'.
+#
+# TODO: should we provide a __post_init__ function like dataclass does?
+
+def _mk_init_fn( self_name, fields ):
+
+  # Sanity check: make sure no postional argument after keyword argument
+  seen_kwarg = False
+  for f in fields:
+    if not f.default is MISSING:
+      seen_kwarg = True
+    elif seen_kwarg:
+      raise TypeError( f'Non keyword argument {f.name} '
+                       'follows keyword argument' )
+
+  _globals = { 'MISSING' : MISSING }
+
+  # Register default values to _globals
+  for f in fields:
+    if not (f.default is MISSING):
+      _globals[f'_deflt_{f.name}'] = f.default
+
+  # Register types in _locals
+  _locals = { f'_type_{f.name}' : f.type_ for f in fields }
+
+  return _create_fn(
+    '__init__',
+    [ self_name ] + [ _mk_init_arg( f ) for f in fields ],
+    [ _mk_init_body( self_name, f ) for f in fields ],
+    _locals  = _locals,
+    _globals = _globals,
+  )
+
+#-------------------------------------------------------------------------
+# _mk_str_fn
+#-------------------------------------------------------------------------
+# Creates a __str__ function based on fields. For example, if fields
+# contains two field x (Bits4) and y (Bits4), _mk_init_fn will return a
+# function that looks like the following:
+#
+# def __str__( self ):
+#   return f'{self.x}:{self.y}'
+
+def _mk_str_fn( fields ):
+  return _create_fn(
+    '__str__',
+    [ 'self' ],
+    [ 'return f"' +
+      ':'.join([ f'{{self.{f.name}}}' for f in fields ]) + '"']
+  )
+
+#-------------------------------------------------------------------------
+# _mk_repr_fn
+#-------------------------------------------------------------------------
+# Creates a __repr__ function based on fields. For example, if fields
+# contains two field x (Bits4) and y (Bits4), _mk_init_fn will return a
+# function that looks like the following:
+#
+# def __repr__( self ):
+#  return self.__class__.qualname__ + f'(x={self.x!r}, y={self.y!r})'
+
+def _mk_repr_fn( fields ):
+  return _create_fn(
+    '__repr__',
+    [ 'self' ],
+    [ 'return self.__class__.__qualname__ + f"(' +
+      ', '.join([ f'{f.name}={{self.{f.name}!r}}' for f in fields ]) +
+      ')"']
+  )
+
+#-------------------------------------------------------------------------
+# _mk_eq_fn
+#-------------------------------------------------------------------------
+# Creates a __eq__ function based on fields. By default it just compares
+# each field. For example, if fields contains two field x (Bits4) and y
+# (Bits4), _mk_init_fn will return a function that looks like the
+# following:
+#
+# def __eq__( self, other ):
+#   if other.__class__ is self.__class__:
+#     return (self.x,self.y,) == (other.x,other.y,)
+#   else:
```

Comment:
I wonder if every struct comparison needs to call  .__class__. If so, it might hurt performance. Maybe you could use try-catch block to do this, something like:
```
try:
     return (self.x,self.y,) == (other.x,other.y,)
except Exception as e:
    <handle the exception -- print struct comparison failure>
```

Reply:
hmm that way we are allowing comparison between different types? Than may cause Pymtl behaves differently from Verilator though

Reply:
On my end, I think the if statement and comparisons involve __class__ will kill the JIT...

Reply:
@ptpan @yodada nominal vs structural typing gurus please chime in here

Reply:
i don't think we ever directly compare two bit structs except in the test harness... i think the default `__eq__` should be conservative and should not behave differently from that translated verilog. Besides, user can just overwrites the default `__eq__` if they are not happy with it